core/services/vrf/v2/listener_v2_log_listener.go

```diff
@@ -0,0 +1,451 @@
+		[]common.Hash{lsn.job.VRFSpec.PublicKey.MustHash()}, // topic values
+		fromTimestamp,       // from time
+		logpoller.Finalized, // confs
+	)
+	if err != nil {
+		return 0, errors.Wrap(err, "LogPoller.LogsCreatedAfter RandomWordsRequested logs")
+	}
+
+	// fulfillments don't have keyhash indexed, we'll have to get all of them
+	// TODO: can we instead write a single query that joins on request id's somehow?
+	fulfillments, err := lp.LogsCreatedAfter(
+		lsn.coordinator.RandomWordsFulfilledTopic(), // event sig
+		lsn.coordinator.Address(),                   // address
+		fromTimestamp,                               // from time
+		logpoller.Finalized,                         // confs
+	)
+	if err != nil {
+		return 0, errors.Wrap(err, "LogPoller.LogsCreatedAfter RandomWordsFulfilled logs")
+	}
+
+	unfulfilled, _, _ := lsn.getUnfulfilled(append(requests, fulfillments...), ll)
+	var earliestUnfulfilledBlock int64 = math.MaxInt64
+	for _, req := range unfulfilled {
+		if req.Raw().BlockNumber < uint64(earliestUnfulfilledBlock) {
+			earliestUnfulfilledBlock = int64(req.Raw().BlockNumber)
+		}
+	}
+	if earliestUnfulfilledBlock == math.MaxInt64 {
+		// no unfulfilled requests
+		return latestBlock.FinalizedBlockNumber, nil
+	}
+
+	// earliestUnfulfilledBlock is <= latestBlock.FinalizedBlockNumber because in our queries we specify
+	// logpoller.Finalized as the number of confirmations, so it's impossible for this not to be the case.
+	return earliestUnfulfilledBlock, nil
+}
+
+func (lsn *listenerV2) updateLastProcessedBlock(ctx context.Context, currLastProcessedBlock int64) (lastProcessedBlock int64, err error) {
+	lp := lsn.chain.LogPoller()
+	start := time.Now()
+
+	latestBlock, err := lp.LatestBlock(pg.WithParentCtx(ctx))
+	if err != nil {
+		lsn.l.Errorw("error getting latest block", "err", err)
+		return 0, errors.Wrap(err, "LogPoller.LatestBlock()")
+	}
+	ll := lsn.l.With(
+		"currLastProcessedBlock", currLastProcessedBlock,
+		"latestBlock", latestBlock.BlockNumber,
+		"latestFinalizedBlock", latestBlock.FinalizedBlockNumber)
+	ll.Debugw("updating last processed block")
+	defer func() {
+		ll.Debugw("done updating last processed block", "elapsed", time.Since(start))
+	}()
+
+	logs, err := lp.LogsWithSigs(
+		currLastProcessedBlock,
+		latestBlock.FinalizedBlockNumber,
+		[]common.Hash{lsn.coordinator.RandomWordsFulfilledTopic(), lsn.coordinator.RandomWordsRequestedTopic()},
+		lsn.coordinator.Address(),
+		pg.WithParentCtx(ctx),
+	)
+	if err != nil {
+		return currLastProcessedBlock, errors.Wrap(err, "LogPoller.LogsWithSigs")
+	}
+
+	unfulfilled, _, _ := lsn.getUnfulfilled(logs, ll)
+	// find request block of earliest unfulfilled request
+	var earliestUnprocessedRequestBlock int64 = math.MaxInt64
+	for _, req := range unfulfilled {
+		if req.Raw().BlockNumber < uint64(earliestUnprocessedRequestBlock) {
+			earliestUnprocessedRequestBlock = int64(req.Raw().BlockNumber)
+		}
+	}
+
+	// cases:
```

Comment:
can this be simplified in the same way as previous comment? initialize to latestBlock.Finalized... and let the for loop find anything that is smaller?